[Sudo rules][Settings] Implement the 'Who' section

carma12 opened the pull request:
The 'Who' section allows to create and associate Users (already-defined or externals) and User groups to a given Sudo rule. It is also possible to apply the association to specific users or to anyone.

**Disclaimer:** When selecting the `Anyone` radio, the 'Save'  button is enabled. Once it's clicked, two API calls are called: one to delete the existing users and user groups, and the other to modify the current sudo rule. This functionality requires to re-implement the way the changes are detected by the `ipaObject` and perform the corresponding IPA calls, thus this functionality will not work in this PR and will be implemented in a different one.

carma12 commented on src/components/layouts/DualListLayout.tsx:
```diff
@@ -11,16 +18,19 @@ import ExclamationTriangleIcon from "@patternfly/react-icons/dist/esm/icons/excl
+  | "netgroup"
+  | "hbacsvc"
+  | "hbacsvcgroup"
+  | "sudo rule";
```
You are right. Although this specific type is not being used by the `SudoRulesWho` component (it only needs to display `users` and `groups`) I decided to add it in case it might be useful in the future. I also corrected the `getEntries` endpoint from `rpc.ts` file.

carma12 commented:
@mreynolds389 - The code has been corrected based on the comments. Feel free to keep reviewing.

mreynolds389 commented:
Adding/removing users is not updating the Who table correctly.  I added a user, the tab counter was right for users, but the table remained empty, and it showed I added a user group when I did not.  Deleting a user is also not update the table.

carma12 commented:
> @mreynolds389 - The code has been corrected based on the comments. Feel free to keep reviewing.

You are right. Actually, I didn't detected this because addition and deletion are working fine when there are elements in the table already. Weird... I will check.

carma12 commented:
> Adding/removing users is not updating the Who table correctly. I added a user, the tab counter was right for users, but the table remained empty, and it showed I added a user group when I did not. Deleting a user is also not updating the table.

You are right. Actually, I didn't detect this because addition and deletion seems to work fine when there are elements in the table already. Weird... I will check.

carma12 commented:
> > Adding/removing users is not updating the Who table correctly. I added a user, the tab counter was right for users, but the table remained empty, and it showed I added a user group when I did not. Deleting a user is also not updating the table.
> 
> You are right. Actually, I didn't detect this because addition and deletion seems to work fine when there are elements in the table already. Weird... I will check.

This has been fixed. Feel free to continue the review.

mreynolds389 commented:
LGTM, but I do have a suggestion/thought.  I think I prefer how the old UI works when handling the "Add external" sudo rules...

In the old UI there was no "Add external" button, instead you just used the add arrow ">" button from the dual select list which then moves the external value over to the chosen options list.  WDYT?

carma12 commented:
> LGTM, but I do have a suggestion/thought. I think I prefer how the old UI works when handling the "Add external" sudo rules...
> 
> In the old UI there was no "Add external" button, instead you just used the add arrow ">" button from the dual select list which then moves the external value over to the chosen options list. WDYT?

I tried to implement it as in the old WebUI at first - But the [PatternFly component](https://www.patternfly.org/components/dual-list-selector/react-next/#basic) doesn't provide a way to introduce a custom variable when this is not defined. I.e., it will require a mechanism to add the element in the same component (like a text input or similar), but this is not possible considering how the PF component is currently implemented. That is why I came up with the solution provided in this PR, I couldn't find any other way to replicate the same behavior. 

From my point of view, we can keep this approach and check with the UX team in case there is a better one to apply here. There may also be some chances that the old design is not optimal either...

mreynolds389 commented:
> > LGTM, but I do have a suggestion/thought. I think I prefer how the old UI works when handling the "Add external" sudo rules...
> > In the old UI there was no "Add external" button, instead you just used the add arrow ">" button from the dual select list which then moves the external value over to the chosen options list. WDYT?
> 
> I tried to implement it as in the old WebUI at first - But the [PatternFly component](https://www.patternfly.org/components/dual-list-selector/react-next/#basic) doesn't provide a way to introduce a custom variable when this is not defined. I.e., it will require a mechanism to add the element in the same component (like a text input or similar), but this is not possible considering how the PF component is currently implemented. That is why I came up with the solution provided in this PR, I couldn't find any other way to replicate the same behavior.
> 
> From my point of view, we can keep this approach and check with the UX team in case there is a better one to apply here. There may also be some chances that the old design is not optimal either...

Ah I see, that's unfortunate.  I just don't like how the Add button it frozen to the bottom of the textInput.  I know why you did it, and inside a dual Select list makes it even harder to make it look nice.  What do you think of this:

![Screenshot from 2024-10-07 11-53-54](https://github.com/user-attachments/assets/a88ba8cc-8138-4c84-9775-4e067f71ef2a)


If you like it here is the diff

```
diff --git a/src/components/layouts/DualListLayout.tsx b/src/components/layouts/DualListLayout.tsx
index 2e0935d..f1f4892 100644
--- a/src/components/layouts/DualListLayout.tsx
+++ b/src/components/layouts/DualListLayout.tsx
@@ -1,7 +1,8 @@
 import React, { ReactNode, useEffect, useState } from "react";
 // PatternFly
 import {
-  ActionGroup,
+  Flex,
+  FlexItem,
   Button,
   DualListSelector,
   FormGroup,
@@ -16,6 +17,7 @@ import InfoCircleIcon from "@patternfly/react-icons/dist/esm/icons/info-circle-i
 import ExclamationTriangleIcon from "@patternfly/react-icons/dist/esm/icons/exclamation-triangle-icon";
 // RPC client
 import { useGetIDListMutation, GenericPayload } from "src/services/rpc";
+import ArrowRightIcon from '@patternfly/react-icons/dist/esm/icons/arrow-right-icon';
 
 export type DualListTarget =
   | "user"
@@ -229,20 +231,31 @@ const DualListTableLayout = (props: PropsToAddModal) => {
             fieldId="dual-list-external"
             style={{ width: "45%" }}
           >
-            <TextInput
-              type="text"
-              id="dual-list-external"
-              name="dual-list-external"
-              aria-label="dual list external"
-              value={externalValue}
-              onChange={(_event, value) => setExternalValue(value)}
-            />
+            <Flex>
+              <FlexItem>
+                <TextInput
+                  type="text"
+                  id="dual-list-external"
+                  name="dual-list-external"
+                  aria-label="dual list external"
+                  value={externalValue}
+                  onChange={(_event, value) => setExternalValue(value)}
+                />
+              </FlexItem>
+              <FlexItem>
+                <Button
+                  title="Add external item to the chosen list"
+                  size="sm"
+                  variant="secondary"
+                  icon={<ArrowRightIcon />}
+                  iconPosition="right"
+                  onClick={onAddExternalValue}
+                >
+                  Add
+                </Button>
+              </FlexItem>
+            </Flex>
           </FormGroup>
-          <ActionGroup className="pf-v5-u-mt-0">
-            <SecondaryButton onClickHandler={onAddExternalValue}>
-              Add external
-            </SecondaryButton>
-          </ActionGroup>
         </>
       ),
     });
```

carma12 commented:
> <Flex>
> +              <FlexItem>
> +                <TextInput
> +                  type="text"
> +                  id="dual-list-external"
> +                  name="dual-list-external"
> +                  aria-label="dual list external"
> +                  value={externalValue}
> +                  onChange={(_event, value) => setExternalValue(value)}
> +                />
> +              </FlexItem>
> +              <FlexItem>
> +                <Button
> +                  title="Add external item to the chosen list"
> +                  size="sm"
> +                  variant="secondary"
> +                  icon={<ArrowRightIcon />}
> +                  iconPosition="right"
> +                  onClick={onAddExternalValue}
> +                  isDisabled={externalValue === ""}
> +                >
> +                  Add
> +                </Button>
> +              </FlexItem>
> +            </Flex>

This definitely looks better! I already changed the code to adapt it. Thanks!

mreynolds389 reviewed (approved):
LGTM!!
